jaxlib/mosaic/gpu/custom_call.cc

```diff
@@ -109,6 +110,106 @@ namespace ffi = xla::ffi;
 using MosaicInitFunc = void(void****);
 using MosaicHostFunc = void(void**);
 
+class TemporaryDirectory {
+ private:
+  TemporaryDirectory(std::string path) : path(std::move(path)) {}
+  // TODO(apaszke): Unlink in destructor.
+
+ public:
+  static absl::StatusOr<TemporaryDirectory> Create() {
+    std::string pattern = "/tmp/mosaic-gpu-XXXXXX";
+    if (mkdtemp(pattern.data()) == NULL) {
+      return absl::InternalError("Failed to create temporary directory");
+    }
+    return TemporaryDirectory(std::move(pattern));
+  }
+
+  std::string_view GetPath() { return path; }
+
+ private:
+  std::string path;
+};
+
+absl::StatusOr<std::string> RunCUDATool(const char* tool,
+                                        const std::vector<const char*>& args,
+                                        bool stderr_to_stdout = true) {
+  CHECK(!args.empty() && args.back() == nullptr);
+  const char * cuda_path_ptr = getenv("CUDA_ROOT");
+  if (!cuda_path_ptr) return absl::InternalError("Failed to get CUDA_ROOT");
+  std::string tool_path(cuda_path_ptr);
```

Comment:
If you have tsl here, you can use `JoinPath`. If not, `absl::StrCat`.

Reply:
I had to move this function up in the file, so confusingly the entire function appears in the diff. I'm thinking to not address these here as out of scope for the PR, but the comments are fair.

Reply:
Yep let's not change too much in a single PR

Reply:
SGTM